velox/exec/tests/utils/Cursor.cpp

```diff
@@ -125,113 +125,242 @@ bool TaskQueue::hasNext() {
-    executor_ = std::make_shared<folly::CPUThreadPoolExecutor>(
-        std::thread::hardware_concurrency());
-    static std::atomic<uint64_t> cursorQueryId{0};
-    queryCtx = std::make_shared<core::QueryCtx>(
-        executor_.get(),
-        core::QueryConfig({}),
-        std::unordered_map<std::string, std::shared_ptr<Config>>{},
-        cache::AsyncDataCache::getInstance(),
-        nullptr,
-        nullptr,
-        fmt::format("TaskCursorQuery_{}", cursorQueryId++));
+class TaskCursorBase : public TaskCursor {
+ public:
+  explicit TaskCursorBase(const CursorParameters& params) {
+    if (params.queryCtx) {
+      queryCtx_ = params.queryCtx;
+    } else {
+      // NOTE: the destructor of 'executor_' will wait for all the async task
+      // activities to finish on TaskCursor destruction.
+      executor_ = std::make_shared<folly::CPUThreadPoolExecutor>(
+          std::thread::hardware_concurrency());
+      static std::atomic<uint64_t> cursorQueryId{0};
+      queryCtx_ = std::make_shared<core::QueryCtx>(
+          executor_.get(),
+          core::QueryConfig({}),
+          std::unordered_map<std::string, std::shared_ptr<Config>>{},
+          cache::AsyncDataCache::getInstance(),
+          nullptr,
+          nullptr,
+          fmt::format("SingleThreadedTaskCursorQuery_{}", cursorQueryId++));
+    }
+
+    if (!params.queryConfigs.empty()) {
+      auto configCopy = params.queryConfigs;
+      queryCtx_->testingOverrideConfigUnsafe(std::move(configCopy));
+    }
+
+    planFragment_ = {
+        params.planNode,
+        params.executionStrategy,
+        params.numSplitGroups,
+        params.groupedExecutionLeafNodeIds};
+    const std::string taskId =
+        fmt::format("test_single_threaded_cursor {}", ++serial_);
+
+    if (!params.spillDirectory.empty()) {
+      taskSpillDirectory_ = params.spillDirectory + "/" + taskId;
+      auto fileSystem =
+          velox::filesystems::getFileSystem(taskSpillDirectory_, nullptr);
+      VELOX_CHECK_NOT_NULL(fileSystem, "File System is null!");
+      try {
+        fileSystem->mkdir(taskSpillDirectory_);
+      } catch (...) {
+        LOG(ERROR) << "Faield to create task spill directory "
+                   << taskSpillDirectory_ << " base director "
+                   << params.spillDirectory << " exists["
+                   << std::filesystem::exists(taskSpillDirectory_) << "]";
+
+        std::rethrow_exception(std::current_exception());
+      }
+
+      LOG(INFO) << "Task spill directory[" << taskSpillDirectory_
+                << "] created";
+    }
   }
 
-  if (!params.queryConfigs.empty()) {
-    auto configCopy = params.queryConfigs;
-    queryCtx->testingOverrideConfigUnsafe(std::move(configCopy));
+ protected:
+  static std::atomic<int32_t> serial_;
+  std::shared_ptr<core::QueryCtx> queryCtx_;
+  core::PlanFragment planFragment_;
+  std::string taskSpillDirectory_;
+
+ private:
+  std::shared_ptr<folly::Executor> executor_;
```

Comment:
We can move executor_ to MultiThreadedTaskCursor?

Reply:
It's nullptr for single-threaded cursor but still used in the base class to instantiate QueryCtx. So let's just keep it here for this patch?